File: create_package.py
#!/usr/bin/env python3
"""
Create Unity Package (.unitypackage) without Unity Editor
Unity packages are tar.gz archives with a specific structure.
"""

import os
import tarfile
import tempfile
import shutil
import re

# Configuration
PACKAGE_NAME = "Deffatest_v1.0.0.unitypackage"
ASSETS_FOLDER = "Assets/Deffatest"

def extract_guid(meta_file_path):
    """Extract GUID from a .meta file"""
    try:
        with open(meta_file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            match = re.search(r'guid:\s*([a-f0-9]+)', content)
            if match:
                return match.group(1)
    except Exception as e:
        print(f"Error reading {meta_file_path}: {e}")
    return None

def get_all_assets(base_path):
    """Get all assets and their meta files"""
    assets = []
    
    for root, dirs, files in os.walk(base_path):
        # Add directory meta files
        dir_meta = root + ".meta"
        if os.path.exists(dir_meta):
            rel_path = root.replace("\\", "/")
            assets.append({
                'path': rel_path,
                'meta': dir_meta,
                'is_folder': True,
                'asset': None
            })
        
        # Add file assets
        for file in files:
            if file.endswith('.meta'):
                continue
            
            file_path = os.path.join(root, file)
            meta_path = file_path + ".meta"
            
            if os.path.exists(meta_path):
                rel_path = file_path.replace("\\", "/")
                assets.append({
                    'path': rel_path,
                    'meta': meta_path,
                    'is_folder': False,
                    'asset': file_path
                })
    
    # Also add the root folder meta
    root_meta = base_path + ".meta"
    if os.path.exists(root_meta):
        assets.insert(0, {
            'path': base_path.replace("\\", "/"),
            'meta': root_meta,
            'is_folder': True,
            'asset': None
        })
    
    return assets

def create_unitypackage(output_path, assets_folder):
    """Create a .unitypackage file"""
    
    print(f"Creating Unity Package: {output_path}")
    print(f"Assets folder: {assets_folder}")
    
    # Get all assets
    assets = get_all_assets(assets_folder)
    print(f"Found {len(assets)} assets")
    
    # Create temp directory for package structure
    temp_dir = tempfile.mkdtemp()
    
    try:
        for asset in assets:
            guid = extract_guid(asset['meta'])
            if not guid:
                print(f"  Skipping (no GUID): {asset['path']}")
                continue
            
            print(f"  Adding: {asset['path']} (GUID: {guid[:8]}...)")
            
            # Create GUID folder
            guid_folder = os.path.join(temp_dir, guid)
            os.makedirs(guid_folder, exist_ok=True)
            
            # Write pathname file
            pathname_file = os.path.join(guid_folder, "pathname")
            with open(pathname_file, 'w', encoding='utf-8') as f:
                f.write(asset['path'])
            
            # Copy meta file as asset.meta
            meta_dest = os.path.join(guid_folder, "asset.meta")
            shutil.copy2(asset['meta'], meta_dest)
            
            # Copy actual asset file (if not a folder)
            if not asset['is_folder'] and asset['asset']:
                asset_dest = os.path.join(guid_folder, "asset")
                shutil.copy2(asset['asset'], asset_dest)
        
        # Create tar.gz archive
        print(f"\nCreating archive...")
        with tarfile.open(output_path, "w:gz") as tar:
            for item in os.listdir(temp_dir):
                item_path = os.path.join(temp_dir, item)
                tar.add(item_path, arcname=item)
        
        print(f"\n✅ Package created successfully!")
        print(f"   Output: {os.path.abspath(output_path)}")
        print(f"   Size: {os.path.getsize(output_path) / 1024:.1f} KB")
        
    finally:
        # Cleanup temp directory
        shutil.rmtree(temp_dir)

if __name__ == "__main__":
    # Change to script directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(script_dir)
    
    # Create package
    create_unitypackage(PACKAGE_NAME, ASSETS_FOLDER)

File: create_package.js
#!/usr/bin/env node
/**
 * Create Unity Package (.unitypackage) without Unity Editor
 * Unity packages are tar.gz archives with a specific structure.
 */

const fs = require('fs');
const path = require('path');
const zlib = require('zlib');
const { execSync } = require('child_process');

const PACKAGE_NAME = 'Deffatest_v1.0.0.unitypackage';
const ASSETS_FOLDER = 'Assets/Deffatest';

// Extract GUID from meta file
function extractGuid(metaFilePath) {
    try {
        const content = fs.readFileSync(metaFilePath, 'utf8');
        const match = content.match(/guid:\s*([a-f0-9]+)/);
        return match ? match[1] : null;
    } catch (e) {
        console.error(`Error reading ${metaFilePath}: ${e.message}`);
        return null;
    }
}

// Get all assets recursively
function getAllAssets(basePath) {
    const assets = [];
    
    function walkDir(dir) {
        const items = fs.readdirSync(dir);
        
        for (const item of items) {
            const fullPath = path.join(dir, item);
            const stat = fs.statSync(fullPath);
            
            if (stat.isDirectory()) {
                // Check for directory meta file
                const dirMeta = fullPath + '.meta';
                if (fs.existsSync(dirMeta)) {
                    assets.push({
                        path: fullPath.replace(/\\/g, '/'),
                        meta: dirMeta,
                        isFolder: true,
                        asset: null
                    });
                }
                walkDir(fullPath);
            } else if (!item.endsWith('.meta')) {
                // Regular file
                const fileMeta = fullPath + '.meta';
                if (fs.existsSync(fileMeta)) {
                    assets.push({
                        path: fullPath.replace(/\\/g, '/'),
                        meta: fileMeta,
                        isFolder: false,
                        asset: fullPath
                    });
                }
            }
        }
    }
    
    // Add root folder meta
    const rootMeta = basePath + '.meta';
    if (fs.existsSync(rootMeta)) {
        assets.push({
            path: basePath.replace(/\\/g, '/'),
            meta: rootMeta,
            isFolder: true,
            asset: null
        });
    }
    
    walkDir(basePath);
    return assets;
}

// Create tar archive manually (simple implementation)
function createTarEntry(name, content, isDir = false) {
    const header = Buffer.alloc(512, 0);
    
    // File name (100 bytes)
    const nameBytes = Buffer.from(name.substring(0, 99));
    nameBytes.copy(header, 0);
    
    // File mode (8 bytes)
    const mode = isDir ? '0000755' : '0000644';
    Buffer.from(mode + ' ').copy(header, 100);
    
    // UID (8 bytes)
    Buffer.from('0000000 ').copy(header, 108);
    
    // GID (8 bytes)
    Buffer.from('0000000 ').copy(header, 116);
    
    // File size (12 bytes)
    const size = content ? content.length : 0;
    const sizeStr = size.toString(8).padStart(11, '0') + ' ';
    Buffer.from(sizeStr).copy(header, 124);
    
    // Modification time (12 bytes)
    const mtime = Math.floor(Date.now() / 1000).toString(8).padStart(11, '0') + ' ';
    Buffer.from(mtime).copy(header, 136);
    
    // Checksum placeholder (8 bytes)
    Buffer.from('        ').copy(header, 148);
    
    // Type flag (1 byte)
    header[156] = isDir ? 53 : 48; // '5' for dir, '0' for file
    
    // Calculate checksum
    let checksum = 0;
    for (let i = 0; i < 512; i++) {
        checksum += header[i];
    }
    const checksumStr = checksum.toString(8).padStart(6, '0') + '\0 ';
    Buffer.from(checksumStr).copy(header, 148);
    
    // Prepare data blocks
    const blocks = [header];
    
    if (content && content.length > 0) {
        // Add content with padding to 512-byte blocks
        const paddingSize = 512 - (content.length % 512);
        const paddedContent = Buffer.concat([
            Buffer.isBuffer(content) ? content : Buffer.from(content),
            Buffer.alloc(paddingSize === 512 ? 0 : paddingSize, 0)
        ]);
        blocks.push(paddedContent);
    }
    
    return Buffer.concat(blocks);
}

function createUnityPackage(outputPath, assetsFolder) {
    console.log(`Creating Unity Package: ${outputPath}`);
    console.log(`Assets folder: ${assetsFolder}`);
    
    const assets = getAllAssets(assetsFolder);
    console.log(`Found ${assets.length} assets\n`);
    
    const tarParts = [];
    
    for (const asset of assets) {
        const guid = extractGuid(asset.meta);
        if (!guid) {
            console.log(`  Skipping (no GUID): ${asset.path}`);
            continue;
        }
        
        console.log(`  Adding: ${asset.path} (GUID: ${guid.substring(0, 8)}...)`);
        
        // Create directory entry
        tarParts.push(createTarEntry(`${guid}/`, null, true));
        
        // Add pathname file
        tarParts.push(createTarEntry(`${guid}/pathname`, asset.path));
        
        // Add asset.meta
        const metaContent = fs.readFileSync(asset.meta);
        tarParts.push(createTarEntry(`${guid}/asset.meta`, metaContent));
        
        // Add asset file (if not a folder)
        if (!asset.isFolder && asset.asset) {
            const assetContent = fs.readFileSync(asset.asset);
            tarParts.push(createTarEntry(`${guid}/asset`, assetContent));
        }
    }
    
    // Add tar end blocks (two 512-byte zero blocks)
    tarParts.push(Buffer.alloc(1024, 0));
    
    // Combine all parts
    const tarBuffer = Buffer.concat(tarParts);
    
    // Gzip compress
    console.log(`\nCompressing archive...`);
    const gzipped = zlib.gzipSync(tarBuffer);
    
    // Write output
    fs.writeFileSync(outputPath, gzipped);
    
    const sizeKB = (fs.statSync(outputPath).size / 1024).toFixed(1);
    console.log(`\n✅ Package created successfully!`);
    console.log(`   Output: ${path.resolve(outputPath)}`);
    console.log(`   Size: ${sizeKB} KB`);
}

// Run
createUnityPackage(PACKAGE_NAME, ASSETS_FOLDER);
